pkg/awsutils/awsutils.go

```diff
@@ -513,75 +526,246 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata(ctx context.Context)
 	return nil
 }
 
-// RefreshSGIDs retrieves security groups
-func (cache *EC2InstanceMetadataCache) RefreshSGIDs(mac string, dsAccess *datastore.DataStoreAccess) error {
-	ctx := context.TODO()
+// discoverCustomSecurityGroups discovers security groups with the cni-role tag
+func (cache *EC2InstanceMetadataCache) discoverCustomSecurityGroups(ctx context.Context) ([]string, error) {
+	describeSGInput := &ec2.DescribeSecurityGroupsInput{
+		Filters: []ec2types.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []string{cache.vpcID},
+			},
+			{
+				Name:   aws.String("tag:" + subnetDiscoveryTagKey),
+				Values: []string{subnetDiscoveryTagValueIncluded},
+			},
+		},
+	}
 
-	sgIDs, err := cache.imds.GetSecurityGroupIDs(ctx, mac)
+	result, err := cache.ec2SVC.DescribeSecurityGroups(ctx, describeSGInput)
 	if err != nil {
-		awsAPIErrInc("GetSecurityGroupIDs", err)
-		return err
+		return nil, fmt.Errorf("discoverCustomSecurityGroups: unable to describe security groups: %v", err)
+	}
+
+	var sgIDs []string
+	for _, sg := range result.SecurityGroups {
+		sgIDs = append(sgIDs, *sg.GroupId)
 	}
 
-	newSGs := StringSet{}
-	newSGs.Set(sgIDs)
-	addedSGs := newSGs.Difference(&cache.securityGroups)
-	addedSGsCount := 0
-	deletedSGs := cache.securityGroups.Difference(&newSGs)
-	deletedSGsCount := 0
+	return sgIDs, nil
+}
+
+// getENISubnetID gets the subnet ID for an ENI from AWS
+func (cache *EC2InstanceMetadataCache) getENISubnetID(ctx context.Context, eniID string) (string, error) {
+	describeInput := &ec2.DescribeNetworkInterfacesInput{
+		NetworkInterfaceIds: []string{eniID},
+	}
+
+	result, err := cache.ec2SVC.DescribeNetworkInterfaces(ctx, describeInput)
+	if err != nil {
+		return "", fmt.Errorf("getENISubnetID: unable to describe network interface: %v", err)
+	}
+
+	if len(result.NetworkInterfaces) == 0 {
+		return "", fmt.Errorf("getENISubnetID: no interfaces found")
+	}
+
+	return *result.NetworkInterfaces[0].SubnetId, nil
+}
+
+// Helper function to check if an ENI is in a secondary subnet
+func (cache *EC2InstanceMetadataCache) isENIInSecondarySubnet(ctx context.Context, eniID string) bool {
+	eniSubnetID, err := cache.getENISubnetID(ctx, eniID)
+	return err == nil && eniSubnetID != cache.subnetID
```

Comment:
nit: considering change cache.subnetID to cache.primarySubnetID or cache.nodeSubnetID.